File: src/content/docs/topics/cs/leetcode-150/arrays-and-hashing/049-group-anagrams.py
def group_anagrams(strs: list[str]) -> list[list[str]]:
    pass  # TODO: implement

def _run_tests():

    def normalize(result):
        return sorted((sorted(g) for g in result))
    r1 = group_anagrams(['eat', 'tea', 'tan', 'ate', 'nat', 'bat'])
    assert normalize(r1) == [['ate', 'eat', 'tea'], ['bat'], ['nat', 'tan']]
    r2 = group_anagrams([''])
    assert normalize(r2) == [['']]
    r3 = group_anagrams(['a'])
    assert normalize(r3) == [['a']]
    r4 = group_anagrams(['abc', 'bca', 'cab'])
    assert normalize(r4) == [['abc', 'bca', 'cab']]
    r5 = group_anagrams(['a', 'b', 'c'])
    assert normalize(r5) == [['a'], ['b'], ['c']]
    print('all tests pass')

if __name__ == '__main__':
    _run_tests()

File: src/content/docs/topics/cs/coding-problems/arrays-and-hashing/049-group-anagrams.ts
function assert(condition: boolean, msg: string = ''): void {
    if (!condition) throw new Error(msg || 'Assertion failed');
}

function normalize(result: string[][]): string[][] {
    return result.map(g => [...g].sort()).sort((a, b) => a[0] < b[0] ? -1 : a[0] > b[0] ? 1 : 0);
}

function groupAnagrams(strs: string[]): string[][] {
    // TODO: implement
    return [];
}

function _runTests(): void {
    const r1 = groupAnagrams(['eat', 'tea', 'tan', 'ate', 'nat', 'bat']);
    assert(JSON.stringify(normalize(r1)) === JSON.stringify([['ate', 'eat', 'tea'], ['bat'], ['nat', 'tan']]));
    const r2 = groupAnagrams(['']);
    assert(JSON.stringify(normalize(r2)) === JSON.stringify([['']]));
    const r3 = groupAnagrams(['a']);
    assert(JSON.stringify(normalize(r3)) === JSON.stringify([['a']]));
    const r4 = groupAnagrams(['abc', 'bca', 'cab']);
    assert(JSON.stringify(normalize(r4)) === JSON.stringify([['abc', 'bca', 'cab']]));
    const r5 = groupAnagrams(['a', 'b', 'c']);
    assert(JSON.stringify(normalize(r5)) === JSON.stringify([['a'], ['b'], ['c']]));
    // perf
    const big = Array.from({ length: 500_000 }, (_, i) => 'a'.repeat((i % 5) + 1));
    const t0 = performance.now();
    groupAnagrams(big);
    console.log(`perf groupAnagrams n=500000 strings: ${(performance.now() - t0).toFixed(1)}ms`);
    console.log('all tests pass');
}

_runTests();
